File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/media/util/RemoteMediaSecurity.java
package crabcraft.net.crabUtilities.media.util;

import crabcraft.net.crabUtilities.media.source.MediaSourceKind;

import java.util.function.Predicate;

/** Shared sink-time checks for remote disc and horn creation and playback. */
public final class RemoteMediaSecurity {
  public static final float MIN_VOLUME = 0f;
  public static final float MAX_VOLUME = 2f;

  private RemoteMediaSecurity() {}

  public static boolean canCreate(Predicate<String> hasPermission, MediaSourceKind sourceKind) {
    return sourceKind != null
      && hasPermission.test("crabutilities.media.create")
      && hasPermission.test("crabutilities.media.create.remote")
      && hasPermission.test("crabutilities.media.create.remote." + sourceKind.permissionSuffix());
  }

  public static boolean isValidVolume(float volume) {
    return Float.isFinite(volume) && volume >= MIN_VOLUME && volume <= MAX_VOLUME;
  }

  public static boolean isValidDiscSettings(float volume, int distance, int rangeMin, int rangeMax) {
    return isValidVolume(volume)
      && rangeMin >= 0
      && rangeMax >= rangeMin
      && distance >= rangeMin
      && distance <= rangeMax;
  }

  public static int playbackDistance(int storedDistance, int configuredDefault,
                                     int rangeMin, int rangeMax) {
    int selected = storedDistance > 0 ? storedDistance : configuredDefault;
    return Math.max(rangeMin, Math.min(rangeMax, selected));
  }
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/media/util/RemoteMediaSecurity.kt
package crabcraft.net.crabUtilities.media.util

import crabcraft.net.crabUtilities.media.source.MediaSourceKind
import java.util.function.Predicate

/** Shared sink-time checks for remote disc and horn creation and playback. */
class RemoteMediaSecurity private constructor() {
  companion object {
    const val MIN_VOLUME = 0f
    const val MAX_VOLUME = 2f

    @JvmStatic fun canCreate(hasPermission: Predicate<String>, sourceKind: MediaSourceKind?): Boolean =
      sourceKind != null && hasPermission.test("crabutilities.media.create")
        && hasPermission.test("crabutilities.media.create.remote")
        && hasPermission.test("crabutilities.media.create.remote." + sourceKind.permissionSuffix())

    @JvmStatic fun isValidVolume(volume: Float): Boolean =
      volume.isFinite() && volume >= MIN_VOLUME && volume <= MAX_VOLUME

    @JvmStatic fun isValidDiscSettings(volume: Float, distance: Int, rangeMin: Int, rangeMax: Int): Boolean =
      isValidVolume(volume) && rangeMin >= 0 && rangeMax >= rangeMin
        && distance >= rangeMin && distance <= rangeMax

    @JvmStatic fun playbackDistance(storedDistance: Int, configuredDefault: Int, rangeMin: Int, rangeMax: Int): Int {
      val selected = if (storedDistance > 0) storedDistance else configuredDefault
      return Math.max(rangeMin, Math.min(rangeMax, selected))
    }
  }
}
